### PR TITLE
notmuch: add vim plugin

### PR DESCRIPTION
**Pull request by Julow**
Hi !

This PR adds notmuch's Vim plugin and Ruby bindings. The Ruby bindings are required by the Vim plugin.
The first commit is a bit unrelated, it removes `emacs` from the build dependencies when `withEmacs` is `false`. I don't even want this on my computer ! (kidding, it was just an easy fix)

The Vim plugin is enabled by default because it consists only of small static files. The ruby bindings are not big either and ruby is already added as a dependency. Also emacs is enabled by default, so it's fair.

The way I hook the ruby bindings doesn't seem right. I couldn't find examples in nixpkgs and I think `bundlerEnv` cannot be used here. Is here a better way ?

The ruby bindings have a dependency on the gem `mail` but it is optional and loaded at runtime, so I haven't included it.

There is a related issue: https://github.com/NixOS/nixpkgs/issues/43965

###### Motivation for this change

Make the Vim plugin available from nixpkgs and to use it like this:

```nix
let
  notmuch = (pkgs.notmuch.override {
    withEmacs = false;
  }).overrideAttrs (o: {
    doCheck = false; # Too slow
  });

  # Shadows vim, loading the notmuch plugin and a local vimrc
  notmuch_vim = { vimrc }:
    pkgs.runCommand "notmuch-wrapped-vim" {
      buildInputs = [ notmuch ];
    } ''
      # No absolute path to vim on purpose
      mkdir -p "$out/bin"
      cat <<EOF > "$out/bin/vim"
      #!/usr/bin/env bash
      this="$out"
      PATH=\''${PATH//\$this/} # Remove this package from the PATH
      exec vim -c "set packpath+=\$this/vim" -c "packadd notmuch" -c "source ${vimrc}" "\$@"
      EOF
      chmod +x "$out/bin/vim"
      plugin_out="$out/vim/pack/plugins/opt" # Vim plugins directory structure
      mkdir -p "$plugin_out"
      ln -sf "${notmuch}/share/vim" "$plugin_out/notmuch"
    '';
```

I got a bit creative here, I'll probably use something saner :D

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by kisik21** [comment]
In case I do want to use the "mail" gem, how to enable it?

**Comment by SuperSandro2000** [comment]
Can you update the title to something like ``notmuch: add vim plugin``?

**Comment by Julow** [comment]
Sorry for the broken title.

@kisik21 The mail gem is in nixpkgs already: `pkgs.rubyPackages.mail` and can be added to your user env or to a shell's `buildInputs`.

**Comment by kisik21** [comment]
Oh, so vim's ruby support just takes packages from the environment? Sounds impure but can be taken care of with a wrapper, so not a bug per se, just an inconvenience (or a convenient feature, depending on your point of view)

**Comment by Julow** [comment]
It's searching in paths (`RUBYLIB`) so indeed it's impure but wrapping solves that entirely.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review which is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 104407` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>python27Packages.notmuch2: log <a href="https://termbin.com/aql3">https://termbin.com/aql3</a></li>
  </ul>
</details>
<details>
<summary>
12 packages built:
</summary>
<ul>
<li>aerc</li>
<li>afew</li>
<li>lieer</li>
<li>muchsync</li>
<li>neomutt</li>
<li>notmuch</li>
<li>notmuch-addrlookup</li>
<li>python27Packages.notmuch</li>
<li>python37Packages.notmuch</li>
<li>python37Packages.notmuch2</li>
<li>python38Packages.notmuch</li>
<li>python38Packages.notmuch2</li>
</ul>
</details>

File: pkgs/applications/networking/mailreaders/notmuch/default.nix [hunk 1]
```diff
@@ -90,6 +91,8 @@ stdenv.mkDerivation rec {
 
   postInstall = stdenv.lib.optionalString withEmacs ''
     moveToOutput bin/notmuch-emacs-mua $emacs
+  '' + optionalString withVim ''
+    make -C vim DESTDIR="$out/share/vim" prefix="" install
```
Comment (doronbehar) [inline pkgs/applications/networking/mailreaders/notmuch/default.nix:95]: Is this path working for you? I am fairly sure that it'd be safer to do the following:

```sh
make -C vim DESTDIR="$out/share/vim-plugins/notmuch" prefix="" install
```

Also for Neovim users, it'd be nice if you'd also add here:

```sh
mkdir -p $out/share/nvim
ln -s $out/share/vim-plugins/notmuch $out/share/nvim/site
```

File: pkgs/applications/networking/mailreaders/notmuch/default.nix [hunk 2]
```diff
@@ -93,6 +93,16 @@ stdenv.mkDerivation rec {
     moveToOutput bin/notmuch-emacs-mua $emacs
   '' + optionalString withVim ''
     make -C vim DESTDIR="$out/share/vim" prefix="" install
+  '' + optionalString (!isNull ruby) ''
```
Comment (doronbehar) [inline pkgs/applications/networking/mailreaders/notmuch/default.nix:96]: How about adding a bit more discoverable argument such as `withRuby`?

File: pkgs/applications/networking/mailreaders/notmuch/default.nix [hunk 3]
```diff
@@ -93,6 +93,16 @@ stdenv.mkDerivation rec {
...
+    make -C bindings/ruby install DESTDIR="" prefix="$out" install
+    # This is very specific, see pkgs/development/interpreters/ruby/default.nix
+    mkdir -p $out/nix-support
+    cat > $out/nix-support/setup-hook <<EOF
```
Comment (doronbehar) [inline pkgs/applications/networking/mailreaders/notmuch/default.nix:100]: Could you explain a bit more how does this work? Perhaps you can alternatively patch `notmuch.vim` to use a ruby interpreter that has the bindings from `bindings/ruby`.

**Review by doronbehar — CHANGES_REQUESTED** [review]

**Comment by doronbehar** [comment]
Alternative is in https://github.com/NixOS/nixpkgs/pull/308318 .

**Pull request closed without merging**